From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Manish Bhardwaj <mbhardwaj@nvidia.com>
Date: Thu, 10 Jul 2025 13:54:50 +0000
Subject: [PATCH 098/241] NVIDIA: SAUCE: tegra_hv: add support of async error
 diagnostics in tegra_hv driver

Bug 5401290
Jira ESLC-8700

Change-Id: I7baf7437b91aeca957083a910d73ea1db24da1e6
Signed-off-by: Manish Bhardwaj <mbhardwaj@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3427618
Reviewed-by: Suresh Venkatachalam <skathirampat@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
Reviewed-by: svc-mobile-coverity <svc-mobile-coverity@nvidia.com>
(cherry picked from commit 91997a2a5cb7f0a84a1c61ab504abdfc9c09636e)

diff --git a/drivers/virt/tegra/tegra_hv.c b/drivers/virt/tegra/tegra_hv.c
index a0bfee3792247..43d5d8873d8ff 100644
--- a/drivers/virt/tegra/tegra_hv.c
+++ b/drivers/virt/tegra/tegra_hv.c
@@ -103,6 +103,8 @@ struct tegra_hv_data {
 	struct class *hv_class;
 
 	struct device_node *dev;
+
+	uint32_t vcpu_affinity;
 };
 
 /*
@@ -460,6 +462,70 @@ static ssize_t update_vm_op_store(const struct class *class,
 }
 static CLASS_ATTR_WO(update_vm_op);
 
+static long async_err_diagnostics(void *ptr)
+{
+	int32_t ret = 0;
+
+	/*
+	* Virtualization_System will check SMMU, CBB, and MC errors and report to
+	* FSI if any errors are seen. This will be indirectly observed by User code
+	* running on FSI.
+	*/
+	if (!hyp_smmu_diagnostic()) {
+		ERR("hyp_smmu_diagnostic failed\n");
+		ret = -1;
+		goto fail;
+	}
+
+	if (!hyp_cbb_err_diagnostic()) {
+		ERR("hyp_cbb_err_diagnostic failed\n");
+		ret = -1;
+		goto fail;
+	}
+
+	if (!hyp_mc_err_diagnostic()) {
+		ERR("hyp_mc_err_diagnostic failed\n");
+		ret = -1;
+		goto fail;
+	}
+
+fail:
+	return ret;
+}
+
+static ssize_t async_err_diagnostics_store(const struct class *class,
+	const struct class_attribute *attr, const char *buf, size_t len)
+{
+	struct tegra_hv_data *hvd = get_hvd();
+	int val = 0;
+	ssize_t ret = 0;
+
+	ret = kstrtoint(buf, 10, &val);
+	if (ret < 0) {
+		pr_err("%s: Failed to convert string to uint %ld\n", __func__, ret);
+		ret = -EINVAL;
+		goto fail;
+	}
+
+	if (val != 1) {
+		pr_err("%s: Unsupported value, %d\n", __func__, val);
+		ret = -EINVAL;
+		goto fail;
+	}
+
+	if (!cpu_online(hvd->vcpu_affinity)) {
+		pr_warn("CPU %u is not online\n", hvd->vcpu_affinity);
+		ret = -EINVAL;
+		goto fail;
+	}
+
+	ret = work_on_cpu(hvd->vcpu_affinity, async_err_diagnostics, NULL);
+fail:
+	ret = ret ? ret : len;
+	return ret;
+}
+static CLASS_ATTR_WO(async_err_diagnostics);
+
 static int tegra_hv_setup(struct tegra_hv_data *hvd)
 {
 	const uint32_t intr_property_size = 3u;
@@ -502,6 +568,12 @@ static int tegra_hv_setup(struct tegra_hv_data *hvd)
 		}
 	}
 
+	ret = class_create_file(hvd->hv_class, &class_attr_async_err_diagnostics);
+	if (ret != 0) {
+		ERR("failed to create async_err_diagnostics file: %d\n", ret);
+		return ret;
+	}
+
 	ret = hyp_read_ivc_info(&info_page);
 	if (ret != 0) {
 		ERR("failed to obtain IVC info page: %d\n", ret);
@@ -1015,9 +1087,20 @@ void tegra_hv_ivc_channel_reset(struct tegra_hv_ivc_cookie *ivck)
 }
 EXPORT_SYMBOL(tegra_hv_ivc_channel_reset);
 
+static long int read_mpidr(void *data)
+{
+	uint64_t mpidr;
+	__asm volatile("MRS %0, MPIDR_EL1 " : "=r"(mpidr) :: "memory");
+	return mpidr;
+}
+
 static int tegra_hv_probe(struct platform_device *pdev)
 {
 	struct tegra_hv_data *hvd;
+	long int lcpu0_mpidr;
+	long int mpidr = 0UL;
+	uint32_t num_vcpus = num_present_cpus();
+	uint32_t idx;
 	int ret;
 
 	if (!is_tegra_hypervisor_mode())
@@ -1029,6 +1112,20 @@ static int tegra_hv_probe(struct platform_device *pdev)
 		return -ENOMEM;
 	}
 
+	hvd->vcpu_affinity = num_vcpus;
+	lcpu0_mpidr = hyp_lcpu0_mpidr();
+	for (idx = 0; idx < num_vcpus; idx++) {
+		mpidr = work_on_cpu(idx, read_mpidr, NULL);
+		if (mpidr == lcpu0_mpidr) {
+			hvd->vcpu_affinity = idx;
+			break;
+		}
+	}
+
+	if (hvd->vcpu_affinity >= num_vcpus)
+		INFO("%s: cpu affinity (%d) > online cpus (%d)\n", __func__,
+						hvd->vcpu_affinity, num_online_cpus());
+
 	ret = tegra_hv_setup(hvd);
 	if (ret != 0) {
 		tegra_hv_cleanup(hvd);
diff --git a/include/soc/tegra/virt/syscalls.h b/include/soc/tegra/virt/syscalls.h
index c33a01fc6714c..03f97c46976ae 100644
--- a/include/soc/tegra/virt/syscalls.h
+++ b/include/soc/tegra/virt/syscalls.h
@@ -18,9 +18,13 @@
 #define HVC_NR_READ_HYP_INFO		9
 #define HVC_NR_GUEST_RESET		10
 #define HVC_NR_SYSINFO_IPA		13
+#define HVC_NR_LCPU0_MPIDR		15
 #define HVC_NR_READ_VM_INFO		16
-#define HVC_NR_TRACE_GET_EVENT_DATA		0x8003U
-#define HVC_NR_TRACE_SET_EVENT_DATA		0x8004U
+#define HVC_NR_TRACE_GET_EVENT_DATA	0x8003U
+#define HVC_NR_TRACE_SET_EVENT_DATA	0x8004U
+#define HVC_NR_SMMU_DIAG		0x8009U
+#define HVC_NR_CBB_DIAG			0x8133U
+#define HVC_NR_MC_DIAG			0x8134U
 
 #define GUEST_PRIMARY		0
 #define GUEST_IVC_SERVER	0
@@ -579,6 +583,42 @@ static inline int hyp_trace_set_profiler_freq(uint64_t freq)
 	return (int) args[0];
 }
 
+__attribute__((no_sanitize_address))
+static inline bool hyp_smmu_diagnostic(void)
+{
+	uint64_t args[4] = { 0U, 0U, 0U, 0U };
+	hyp_call44(HVC_NR_SMMU_DIAG, args);
+
+	return (args[0] == 0U);
+}
+
+__attribute__((no_sanitize_address))
+static inline bool hyp_cbb_err_diagnostic(void)
+{
+	uint64_t args[4] = { 0U, 0U, 0U, 0U };
+	hyp_call44(HVC_NR_CBB_DIAG, args);
+
+	return (args[0] == 0U);
+}
+
+__attribute__((no_sanitize_address))
+static inline bool hyp_mc_err_diagnostic(void)
+{
+	uint64_t args[4] = { 0U, 0U, 0U, 0U };
+	hyp_call44(HVC_NR_MC_DIAG, args);
+
+	return (args[0] == 0U);
+}
+
+__attribute__((no_sanitize_address))
+static inline long int hyp_lcpu0_mpidr(void)
+{
+	uint64_t args[4] = { 0U, 0U, 0U, 0U };
+	hyp_call44(HVC_NR_LCPU0_MPIDR, args);
+
+	return args[0];
+}
+
 #undef _X3_X17
 #undef _X4_X17
 
-- 
2.43.0

